File: rename_users.py
import os
import sys
import asyncio
import discord
from dotenv import load_dotenv

load_dotenv()

TOKEN = os.getenv("DISCORD_TOKEN")
GUILD_ID_STR = os.getenv("DISCORD_GUILD_ID")

if not TOKEN or not GUILD_ID_STR:
    print("Error: DISCORD_TOKEN or DISCORD_GUILD_ID is not set in .env file.")
    sys.exit(1)

try:
    GUILD_ID = int(GUILD_ID_STR)
except ValueError:
    print(f"Error: DISCORD_GUILD_ID '{GUILD_ID_STR}' is not a valid integer.")
    sys.exit(1)

intents = discord.Intents.default()
intents.members = True

class RenameBot(discord.Client):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.filename = os.getenv("LIST_FILE", "list.txt")
        self.mappings = {}

    def get_manual_input(self):
        print("\n--- Ввод данных вручную ---")
        print("Вводите никнейм и ID через пробел (например: PlayerName 123456789012345).")
        print("Для завершения ввода просто нажмите Enter на пустой строке.\n")

        count = 0
        while True:
            count += 1
            try:
                val = input(f"[{count}] Ник и ID: ").strip()
            except EOFError:
                break
            if not val:
                break

            parts = val.split()
            if len(parts) < 2:
                print("Warning: Некорректный формат. Нужно ввести ник и ID через пробел.")
                count -= 1
                continue

            discord_id_str = parts[-1]
            nickname = " ".join(parts[:-1])

            if not discord_id_str.isdigit():
                print(f"Warning: Некорректный Discord ID: '{discord_id_str}'. Он должен состоять только из цифр.")
                count -= 1
                continue

            self.mappings[int(discord_id_str)] = nickname
        
        print(f"\nВведено вручную {len(self.mappings)} участников.")
        return True

    def load_mappings(self):
        print("Выберите источник данных:")
        print("1. Загрузить из файла")
        print("2. Ввести вручную прямо сейчас")
        try:
            choice = input("Введите 1 или 2: ").strip()
        except EOFError:
            choice = "1"

        if choice == "2":
            return self.get_manual_input()

        if not os.path.exists(self.filename):
            print(f"Error: File {self.filename} not found.")
            return False

        with open(self.filename, "r", encoding="utf-8") as f:
            for line_no, line in enumerate(f, 1):
                line = line.strip()
                if not line:
                    continue
                parts = line.split()
                if len(parts) < 2:
                    print(f"Warning: Line {line_no} is malformed: '{line}'")
                    continue
                
                discord_id_str = parts[-1]
                nickname = " ".join(parts[:-1])
                
                if not discord_id_str.isdigit():
                    print(f"Warning: Line {line_no} has an invalid Discord ID: '{discord_id_str}'")
                    continue
                
                self.mappings[int(discord_id_str)] = nickname
        
        print(f"Loaded {len(self.mappings)} mappings from {self.filename}.")
        return True

    async def on_ready(self):
        print(f"Logged in as {self.user} (ID: {self.user.id})")
        print("----------------------------------------")
        
        if not self.load_mappings():
            await self.close()
            return

        guild = self.get_guild(GUILD_ID)
        if not guild:
            print(f"Error: Guild with ID {GUILD_ID} not found.")
            await self.close()
            return
        
        print(f"Connected to: {guild.name} ({guild.id})")
        print("Fetching members...")
        
        try:
            members = {}
            async for member in guild.fetch_members(limit=None):
                members[member.id] = member
            print(f"Cached {len(members)} members.")
        except Exception as e:
            print(f"Failed to fetch members: {e}")
            members = None

        print("Updating nicknames...")
        print("----------------------------------------")
        
        success_count = 0
        skipped_count = 0
        not_found_count = 0
        error_count = 0
        not_found_members = []

        for discord_id, target_nickname in self.mappings.items():
            member = None
            if members is not None:
                member = members.get(discord_id)
            else:
                try:
                    member = await guild.fetch_member(discord_id)
                except discord.NotFound:
                    member = None
                except discord.HTTPException as e:
                    print(f"[ERROR] Fetching member {discord_id}: {e}")
                    error_count += 1
                    continue
            
            if not member:
                print(f"[NOT FOUND] ID {discord_id} ({target_nickname})")
                not_found_members.append({"id": discord_id, "nickname": target_nickname})
                not_found_count += 1
                continue

            current_nick = member.nick if member.nick else member.name
            if current_nick == target_nickname:
                print(f"[SKIP] {member.name} already has nickname '{target_nickname}'")
                skipped_count += 1
                continue

            try:
                print(f"Renaming {member.name} -> '{target_nickname}'...")
                await member.edit(nick=target_nickname)
                print(f"  [SUCCESS] Renamed to '{target_nickname}'")
                success_count += 1
                await asyncio.sleep(1.0)
            except discord.Forbidden:
                print(f"  [ERROR] Cannot change nickname for {member.name}. Bot lacks permission or member's role is higher.")
                error_count += 1
            except discord.HTTPException as e:
                print(f"  [ERROR] Failed to rename {member.name}: {e}")
                error_count += 1
                if e.status == 429:
                    retry_after = getattr(e, 'retry_after', 5.0)
                    print(f"Rate limited. Waiting for {retry_after} seconds...")
                    await asyncio.sleep(retry_after)

        print("----------------------------------------")
        print("Nickname updates completed.")
        print(f"Summary: Success: {success_count}, Skipped: {skipped_count}, Not Found: {not_found_count}, Errors: {error_count}")
        
        if not_found_members:
            print("\n----------------------------------------")
            print("СПИСОК УЧАСТНИКОВ, КОТОРЫХ НЕТ НА СЕРВЕРЕ:")
            for m in not_found_members:
                print(f"{m['nickname']} {m['id']}")
            print("----------------------------------------")
        
        await self.close()

if __name__ == "__main__":
    bot = RenameBot(intents=intents)
    try:
        bot.run(TOKEN)
    except discord.LoginFailure:
        print("Error: Login failed.")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")

File: rename_users.js
const fs = require('fs');
const readline = require('readline');
const { Client, GatewayIntentBits } = require('discord.js');
require('dotenv').config();

const token = process.env.DISCORD_TOKEN;
const guildId = process.env.DISCORD_GUILD_ID;

if (!token || !guildId) {
    console.error("Error: DISCORD_TOKEN or DISCORD_GUILD_ID is not set in .env file.");
    process.exit(1);
}

const client = new Client({
    intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMembers
    ]
});

function askQuestion(query) {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    return new Promise(resolve => rl.question(query, ans => {
        rl.close();
        resolve(ans);
    }));
}

async function getManualInput() {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    const mappings = new Map();
    console.log("\n--- Ввод данных вручную ---");
    console.log("Вводите никнейм и ID через пробел (например: PlayerName 123456789012345).");
    console.log("Для завершения ввода просто нажмите Enter на пустой строке.\n");

    let count = 0;
    while (true) {
        count++;
        const answer = await new Promise(resolve => rl.question(`[${count}] Ник и ID: `, resolve));
        const trimmed = answer.trim();
        if (!trimmed) {
            break;
        }

        const parts = trimmed.split(/\s+/);
        if (parts.length < 2) {
            console.log("Warning: Некорректный формат. Нужно ввести ник и ID через пробел.");
            count--;
            continue;
        }

        const discordId = parts[parts.length - 1];
        const nickname = parts.slice(0, -1).join(' ');

        if (!/^\d+$/.test(discordId)) {
            console.log(`Warning: Некорректный Discord ID: '${discordId}'. Он должен состоять только из цифр.`);
            count--;
            continue;
        }

        mappings.set(discordId, nickname);
    }
    rl.close();
    console.log(`\nВведено вручную ${mappings.size} участников.`);
    return mappings;
}

async function loadMappings() {
    const choice = await askQuestion("Выберите источник данных:\n1. Загрузить из файла\n2. Ввести вручную прямо сейчас\nВведите 1 или 2: ");
    const trimmedChoice = choice.trim();

    if (trimmedChoice === '2') {
        return await getManualInput();
    }

    const filename = process.env.LIST_FILE || 'list.txt';
    const mappings = new Map();

    if (!fs.existsSync(filename)) {
        console.error(`Error: File ${filename} not found in the current directory.`);
        return null;
    }

    const fileStream = fs.createReadStream(filename);
    const rl = readline.createInterface({
        input: fileStream,
        crlfDelay: Infinity
    });

    let lineNo = 0;
    for await (const line of rl) {
        lineNo++;
        const trimmed = line.trim();
        if (!trimmed) continue;

        const parts = trimmed.split(/\s+/);
        if (parts.length < 2) {
            console.warn(`Warning: Line ${lineNo} is malformed: '${trimmed}'`);
            continue;
        }

        const discordId = parts[parts.length - 1];
        const nickname = parts.slice(0, -1).join(' ');

        if (!/^\d+$/.test(discordId)) {
            console.warn(`Warning: Line ${lineNo} has an invalid Discord ID: '${discordId}'`);
            continue;
        }

        mappings.set(discordId, nickname);
    }

    console.log(`Loaded ${mappings.size} mappings from ${filename}.`);
    return mappings;
}

client.once('ready', async () => {
    console.log(`Logged in as ${client.user.tag} (${client.user.id})`);
    console.log("----------------------------------------");

    const mappings = await loadMappings();
    if (!mappings) {
        client.destroy();
        return;
    }

    try {
        const guild = await client.guilds.fetch(guildId);
        if (!guild) {
            console.error(`Error: Guild with ID ${guildId} not found.`);
            client.destroy();
            return;
        }

        console.log(`Connected to: ${guild.name} (${guild.id})`);
        console.log("Fetching members...");
        
        let members;
        try {
            members = await guild.members.fetch();
            console.log(`Cached ${members.size} members.`);
        } catch (e) {
            console.error(`Failed to fetch members: ${e.message}`);
        }

        console.log("Updating nicknames...");
        console.log("----------------------------------------");

        let successCount = 0;
        let skippedCount = 0;
        let notFoundCount = 0;
        let errorCount = 0;
        const notFoundMembers = [];

        for (const [discordId, targetNickname] of mappings.entries()) {
            let member;
            if (members) {
                member = members.get(discordId);
            } else {
                try {
                    member = await guild.members.fetch(discordId);
                } catch (e) {
                    member = null;
                }
            }

            if (!member) {
                console.log(`[NOT FOUND] ID ${discordId} (${targetNickname})`);
                notFoundMembers.push({ id: discordId, nickname: targetNickname });
                notFoundCount++;
                continue;
            }

            if (member.nickname === targetNickname) {
                console.log(`[SKIP] ${member.user.tag} already has nickname '${targetNickname}'`);
                skippedCount++;
                continue;
            }
            try {
                console.log(`Renaming ${member.user.tag} -> '${targetNickname}'...`);
                await member.setNickname(targetNickname);
                console.log(`  [SUCCESS] Renamed to '${targetNickname}'`);
                successCount++;
                await new Promise(resolve => setTimeout(resolve, 1000));
            } catch (e) {
                console.error(`  [ERROR] Failed to rename ${member.user.tag}: ${e.message}`);
                errorCount++;
            }
        }

        console.log("----------------------------------------");
        console.log("Nickname updates completed.");
        console.log(`Summary: Success: ${successCount}, Skipped: ${skippedCount}, Not Found: ${notFoundCount}, Errors: ${errorCount}`);

        if (notFoundMembers.length > 0) {
            console.log("\n----------------------------------------");
            console.log("СПИСОК УЧАСТНИКОВ, КОТОРЫХ НЕТ НА СЕРВЕРЕ:");
            notFoundMembers.forEach(m => {
                console.log(`${m.nickname} ${m.id}`);
            });
            console.log("----------------------------------------");
        }

    } catch (e) {
        console.error(`An error occurred: ${e.message}`);
    } finally {
        client.destroy();
    }
});

client.login(token).catch(err => {
    console.error("Error: Login failed.");
    console.error(err.message);
});
